OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,451 @@
+                    "Uncaught exception processing queue. Error: " + e.GetAllMessages());
+            }
+            finally
+            {
+                _logger.Log(LogLevel.INFO,
+                    "Exiting processing loop. Attempting to flush pending events.");
+                FlushQueue();
+            }
+        }
+
+        /// <summary>
+        /// Immediately send all ODP events in queue
+        /// </summary>
+        public void Flush()
+        {
+            _flushingIntervalDeadline = DateTime.Now.MillisecondsSince1970() +
+                                        (long)_flushInterval.TotalMilliseconds;
+            _eventQueue.Add(_flushSignal);
+        }
+
+        private void FlushQueue()
+        {
+            _flushingIntervalDeadline = DateTime.Now.MillisecondsSince1970() +
+                                        (long)_flushInterval.TotalMilliseconds;
+
+            if (_currentBatch.Count == 0)
+            {
+                return;
+            }
+
+            List<OdpEvent> toProcessBatch;
+            lock (_mutex)
+            {
+                toProcessBatch = new List<OdpEvent>(_currentBatch);
+                _currentBatch.Clear();
+            }
+
+            try
+            {
+                bool shouldRetry;
+                var attemptNumber = 0;
+                do
+                {
+                    shouldRetry = _odpEventApiManager.SendEvents(_odpConfig.ApiKey,
+                        _odpConfig.ApiHost, toProcessBatch);
+                    attemptNumber += 1;
+                } while (shouldRetry && attemptNumber < Constants.MAX_RETRIES);
+            }
+            catch (Exception e)
+            {
+                _errorHandler.HandleError(e);
+                _logger.Log(LogLevel.ERROR, Constants.ODP_SEND_FAILURE_MESSAGE);
+            }
+        }
+
+        /// <summary>
+        /// Stops ODP event processor.
+        /// </summary>
+        public void Stop()
+        {
+            if (Disposed)
+            {
+                _logger.Log(LogLevel.WARN, Constants.ODP_NOT_INTEGRATED_MESSAGE);
+
+                DropQueue();
+
+                return;
+            }
+
+            _eventQueue.Add(_shutdownSignal);
+
+            if (!_executionThread.Join(_timeoutInterval))
+            {
+                _logger.Log(LogLevel.ERROR,
+                    $"Timeout exceeded attempting to close for {_timeoutInterval.Milliseconds} ms");
+            }
+
+            IsStarted = false;
+            _logger.Log(LogLevel.WARN, $"Stopping scheduler.");
+        }
+
+        /// <summary>
+        /// Add event to queue for sending to ODP
+        /// </summary>
+        /// <param name="odpEvent">Event to enqueue</param>
+        public void SendEvent(OdpEvent odpEvent)
+        {
+            if (Disposed || !IsStarted || !_odpConfig.IsReady())
+            {
+                _logger.Log(LogLevel.WARN, Constants.ODP_NOT_INTEGRATED_MESSAGE);
+                return;
+            }
```

Comment:
We need to separate cases. ODP_NOT_INTEGRATED may be a wrong message for other conditions.